### PR TITLE
Auto GEMM tuning with rocBLAS

### PR DESCRIPTION
**Pull request by bpickrel**
Implements a new feature in the rocBLAS API to tune GEMM solutions automatically.  The finalize() method calls a new tune() method which simply iterates through the list of available solutions, timing them all and choosing the fastest.

This PR is incomplete at this time--changes are still in progress.

At the time of creating this PR, these API calls are a Beta-only feature and are not available in regular ROCm releases.  It is enabled with the preprocessor value ```ROCBLAS_BETA_FEATURES_API```, which is set by CMakeLists.txt if the needed rocBLAS version is installed at compile time.

Update: verified that the PR can be built using the Rocm ver 5.5 API and the internal_testing image is no longer needed.  Set the following line in Dockerfile: `RUN sh -c 'echo deb [arch=amd64 trusted=yes] http://repo.radeon.com/rocm/apt/.apt_5.5/ ubuntu main > /etc/apt/sources.list.d/rocm.list'`

See [issue 1553](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/issues/1553)

### CODE REVIEW

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 1]
```diff
@@ -132,6 +133,43 @@ struct rocblas_gemm
...
+        rocblas_initialize();
+        rocblas_status_ check_valid;
+        if(ctx.get_exhaustive_tune_flag() && solution_idx == 0)
+        {
+            if(this->name() == "gpu::gemm")
+            {
+                auto gemmImpl = gemm_impl<float>(
+                    output_shape, input_shapes, alpha, beta, int8_x4_format, compute_fp32);
+                solution_idx = gemmImpl.tune(ctx, input_shapes);
+                check_valid  = gemmImpl.validate(ctx, input_shapes, solution_idx);
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:148]: Only need to validate if we are de-serializing. No need to validate just after tuning since it would be valid for sure.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 2]
```diff
@@ -132,6 +133,43 @@ struct rocblas_gemm
...
+            {
+                auto gemmImpl = gemm_impl<int32_t>(
+                    output_shape, input_shapes, alpha, beta, int8_x4_format, compute_fp32);
+                solution_idx = gemmImpl.tune(ctx, input_shapes);
+                check_valid  = gemmImpl.validate(ctx, input_shapes, solution_idx);
+            }
+        }
+        else if(solution_idx != 0)
+        {
+            auto gemmImpl = gemm_impl<int32_t>(
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:160]: This should check first if it is `quant_gemm` or regular gemm . If it quant-Gemm then only use `int32_t` otherwise use `float`.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 3]
```diff
@@ -132,6 +133,43 @@ struct rocblas_gemm
...
+                output_shape, input_shapes, alpha, beta, int8_x4_format, compute_fp32);
+            check_valid = gemmImpl.validate(ctx, input_shapes, solution_idx);
+        }
+
+        // Check if solution is valid for problem (could be invalid if this model was
+        // tuned with a different rocBLAS version)
+        if(check_valid == rocblas_status_invalid_value)
+        {
+            std::cerr << "WARNING:  tuned solution is invalid; reverting to default" << std::endl;
+            solution_idx = 0;
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:170]: This check is better done inside `validate()`.  `validate()` can return `solution_index` if it is valid else can return `0`, the default.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 4]
```diff
@@ -24,28 +24,492 @@
...
+    bool transa, transb;
+    T alpha, beta;
+    void* alpha_v = nullptr;
+    void* beta_v  = nullptr;
+    rocblas_gemm_flags int8_flag;
+    rocblas_int lda, ldb, ldc, ldd;
+    rocblas_int a_stride, b_stride, c_stride, d_stride;
+    rocblas_datatype compute_type, arg_type, output_type;
+    bool strided_batched = true, is_3inputs = true, compute_fp32 = true;
+    uint32_t flags = 0; //  optional gemm flags.
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:508]: is this used?

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 5]
```diff
@@ -24,28 +24,492 @@
...
+                ctx.finish();
+            };
+            for(rocblas_int hc = 0; hc < hot_calls; ++hc)
+            {
+                ctx.finish();
+                host_time += time<microseconds>(run_func);
+            }
+            // todo:  Measured time dropped from 20 us to about 6.7 us when I raised hot_calls from
+            // 1 to 11. The higher the hot_calls value, the faster per-call time up to at least 25,
+            // and increasing cold_calls makes little or no difference.  Why?
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:480]: Are you still seeing this ? 
Idea of cold calls is to fill up caches.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 6]
```diff
@@ -132,6 +133,43 @@ struct rocblas_gemm
     {
         return shapes.size() - 1;
     }
+
+    void finalize(context& ctx, const shape& output_shape, const std::vector<shape>& input_shapes)
+    {
+        rocblas_initialize();
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:139]: We should call `rocblas_initialize` in get_rocblas in the context.

File: src/targets/gpu/include/migraphx/gpu/rocblas.hpp [hunk 7]
```diff
@@ -25,6 +25,9 @@
 #define MIGRAPHX_GUARD_MIGRAPHLIB_ROCBLAS_HPP
 #include <migraphx/manage_ptr.hpp>
 #include <migraphx/config.hpp>
+// for development:
+#define ROCBLAS_BETA_FEATURES_API
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/rocblas.hpp:29]: We can check if BETA API exists  in  installed rocBLAS or not inside CMake and define ROCBLAS_BETA_FEATURES_API from there. Similar to this : 
https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/a63ee2e0a81aa3a59ff65d7aa67b522993977274/src/targets/gpu/CMakeLists.txt#L221

Comment (umangyadav) [reply src/targets/gpu/include/migraphx/gpu/rocblas.hpp:29]: Since it is BETA api, all the logic related to tuning should be guarded based on whether #ROCBLAS_BETA_FEATURES_API is defined or not.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 1 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp:148]: Can you be more specific? How do we know if we are deserializing, and where in the code?

Comment (umangyadav) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp:148]: After deserializing `finalize()` is called for each operator. 
I mean you can place `validate()`  inside `else_if(solution_idx !=0)`. So that it only validates for the case where tuning is not requested and solution_index is non-default.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 2 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp:160]: Is that exactly the same as the if-else block above, then?  The other kinds of gemm I see defined in README.MD are gpu::int8_gemm_pack_a and gpu::int8_gemm_pack_b -- how to handle them?

Comment (umangyadav) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp:160]: same as earlier if_then_else.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 8]
```diff
@@ -24,28 +24,492 @@
...
+                  << " output_type: " << output_type << "\n";
+        std::cout << "int8_flag: " << int8_flag << "\n";
+    }
+
+    auto create_strided_batched_gemm_get_solutions_args(context& ctx,
+                                                        const std::vector<argument>& args,
+                                                        rocblas_int* list,
+                                                        rocblas_int* list_size) const
+    {
+        return pack(ctx.get_stream().get_rocblas(),
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:356]: You can refactor this using `pack_join()`which can merge two packs. 

e.g you can have function for `create_gemm_args()` which returns common args for running gemm and for calling get_solutions. 

You can then, extend gemm_args using pack_join to make it run with get_solutions.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 9]
```diff
@@ -24,28 +24,492 @@
...
+        }
+        else
+        {
+            auto gemm_args = create_gemm_args(
+                ctx, input_args, rocblas_gemm_flags_check_solution_index, solution_idx);
+            return rocblas_invoke(&rocblas_gemm_ex, gemm_args);
+        }
+    }
+
+    void print_args() const
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:336]: This is mostly for debugging. You can also check if rocBLAS has any environment flags that would dump args for `rocBLAS-bench`. In that  case, we wouldn't need this. Or we can make `print_args()` to print args such that we can directly pass to `rocBLAS-bench`. 
This is just an improvement, You can work on it if you wish after everything else is fixed/done.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 10]
```diff
@@ -24,28 +24,492 @@
...
+            // todo:  Measured time dropped from 20 us to about 6.7 us when I raised hot_calls from
+            // 1 to 11. The higher the hot_calls value, the faster per-call time up to at least 25,
+            // and increasing cold_calls makes little or no difference.  Why?
+            host_time /= hot_calls;
+            std::cout << "Solution index " << sol << ": " << host_time << " us" << std::endl;
+
+            // track current best
+            if(host_time < bestTime)
+            {
+                printf(" current best index  %d\n", sol);
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:487]: This should use `std::cerr` or `std::clog` instead of plain `printf`.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 3 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: done

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 4 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:521]: It's been replaced with individual flag names for the various create...() calls.  I'll hold off on deleting it, though, until I'm sure I'm not going to refactor those yet again.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 10 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:500]: debug code--I don't think we're going to want that line in the delivered product at all.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 8 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:369]: I don't see any examples of similar usage.  Can pack_join put the arguments in the right order?  The different API calls contain different arguments in the middle of the lists, not just concatenated at the end.  Is it possible to insert extra arguments in a specified location in the list?  (And is it even worth it, if it's that much bother?)

Comment (umangyadav) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:369]: pack_join would concatenate two different packs. It can't insert in the middle. 
I was saying we could keep strided_batched_gemm and regular gemm separate, but for each one separately, we could have function `create_common_gemm_args()` and `create_common_strided_batched_gemm_args()` and then we can extend them using pack_join for creating args for either running gemm or for fetching candidate solutions.

File: src/targets/gpu/CMakeLists.txt [hunk 11]
```diff
@@ -235,6 +235,16 @@ else()
     message(STATUS "MIOpen does not have find mode api")
 endif()
 
+check_library_exists(MIOpen "rocblas_gemm_strided_batched_ex" "${MIOPEN_LOCATION}" HAS_ROCBLAS_BETA_FEATURES_API)
```
Comment (umangyadav) [inline src/targets/gpu/CMakeLists.txt:238]: This should check for existence of BETA API e.g. `rocblas_gemm_get_solution()`.

And replace MIOpen with rocBLAS wherever relevant.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 12]
```diff
@@ -168,6 +169,7 @@ struct rocblas_gemm
                 solution_idx = gemmImpl.validate(ctx, input_shapes, solution_idx);
             }
         }
+#endif
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:172]: We also need to consider cases where MIGraphX wasn't compiled with rocBLAS BETA but MXR passed MiGraphX was created with BETA.  In that case, we can't use solution_index other than `0` since it wouldn't be supported by rocBLAS that was linked with MIGraphX during compile time.

File: src/targets/gpu/include/migraphx/gpu/rocblas.hpp [hunk 13]
```diff
@@ -26,8 +26,9 @@
 #include <migraphx/manage_ptr.hpp>
 #include <migraphx/config.hpp>
 // for development:
-#define ROCBLAS_BETA_FEATURES_API
+#ifdef ROCBLAS_BETA_FEATURES_API
 #define ROCBLAS_NO_DEPRECATED_WARNINGS
```
Comment (umangyadav) [inline src/targets/gpu/include/migraphx/gpu/rocblas.hpp:30]: Are you seeing deprecated warnings if you don't define this?

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 1 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: Done.  Refactored finalize() to call either tune() or validate() but not both.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 6 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: done.

File: src/targets/gpu/include/migraphx/gpu/rocblas.hpp (hunk 13 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/rocblas.hpp:30]: The actual warning I see reads like this:  `"rocblas_gemm_ex_get_solutions is a beta feature and is subject to change in future releases"`

File: src/targets/gpu/CMakeLists.txt (hunk 11 above)
Comment (bpickrel) [reply src/targets/gpu/CMakeLists.txt]: This has been fixed and now appears to get it right.

File: src/targets/gpu/include/migraphx/gpu/rocblas.hpp (hunk 13 above)
Comment (umangyadav) [reply src/targets/gpu/include/migraphx/gpu/rocblas.hpp:30]: You can set this inside Cmakelists.txt then along with `ROCBLAS_BETA_FEATURES_API`

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 8 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:325]: ok, done.

File: src/targets/gpu/include/migraphx/gpu/rocblas.hpp (hunk 7 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/rocblas.hpp]: done

Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/rocblas.hpp]: I assume this variable will have to be renamed some day when rocBLAS commits these features to production.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 2 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: done.

File: src/targets/gpu/gemm_impl.cpp [hunk 14]
```diff
@@ -1,280 +0,0 @@
...
-void gemm(context& ctx,
-          const shape& output_shape,
-          const std::vector<argument>& args,
-          int32_t alpha,
-          int32_t beta,
-          bool int8_x4_format,
-          bool compute_fp32)
-{
-    gemm_impl(ctx, output_shape, args, alpha, beta, int8_x4_format, compute_fp32);
-}
```
Comment (pfultz2) [inline src/targets/gpu/gemm_impl.cpp:276]: All this should stay in the .cpp file.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 15]
```diff
@@ -24,28 +24,488 @@
...
+            // Define the function to be timed
+            auto run_func = [&]() {
+                run(ctx, input_args, sol);
+                ctx.finish();
+            };
+            for(rocblas_int hc = 0; hc < hot_calls; ++hc)
+            {
+                ctx.finish();
+                host_time += time<microseconds>(run_func);
+            }
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:471]: This should just use the `time_op` function from the driver. We just need to move it to the gpu lib so it can be reused here.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 9 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:315]: There **are** environment flags like that; see [API_Reference_Guide logging-in-rocblas](https://rocblas.readthedocs.io/en/rocm-5.2.3/API_Reference_Guide.html?highlight=rocBLAS-bench#logging-in-rocblas).  I think that takes care of this with no action required in code.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 16]
```diff
@@ -24,28 +24,502 @@
...
+}
+
+static inline bool is_transposed(const shape& s)
+{
+    return (s.transposed() and s.strides().back() != 1);
+}
+
+static inline rocblas_int get_batch_stride(const shape& a)
+{
+    return a.strides()[a.strides().size() - 3];
```
Comment (bpickrel) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:109]: This is giving an error from a sanitizer, presumably because a shape with less than 3 strides is being passed.  How to handle that situation?  Check and return zero?  Assert?  Is it ever correct for this to receive a lower rank shape?

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 15 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:486]: You'll have to specify what you have in mind for that.  gemm_impl doesn't use instances of the Migraphx operation, which is what time_op takes as an argument.  Instead, we're making the rocblas calls directly.  So we can't reuse it directly--what would we substitute here?  (Also, does this change belong in the same PR?)

**Comment by pfultz2** [comment]
These are the tidy warnings I get using `make tidy-target-migraphx_gpu-gemm_impl_cpp`(which will only run tidy on the gemm_impl.cpp):

```
/data/src/targets/gpu/gemm_impl.cpp:421:16: error: invalid case style for local variable 'bestTime' [readability-identifier-naming,-warnings-as-errors]
        double bestTime   = std::numeric_limits<double>::max();
               ^~~~~~~~
               best_time
/data/src/targets/gpu/gemm_impl.cpp:424:21: error: invalid case style for local variable 'bestSol' [readability-identifier-naming,-warnings-as-errors]
        rocblas_int bestSol = 0;
                    ^~~~~~~
                    best_sol
```

File: src/targets/gpu/CMakeLists.txt [hunk 17]
```diff
@@ -236,6 +240,13 @@ else()
     message(STATUS "MIOpen does not have find mode api")
 endif()
 
+if(HAS_ROCBLAS_BETA_FEATURES_API)
+    target_compile_definitions(migraphx_gpu PUBLIC -DROCBLAS_BETA_FEATURES_API -DROCBLAS_NO_DEPRECATED_WARNINGS)
+    message(STATUS "MIGraphx is using Beta API of MIOpen")
```
Comment (pfultz2) [inline src/targets/gpu/CMakeLists.txt:245]: This is the beta API of rocblas not miopen.

File: src/targets/gpu/gemm_impl.cpp [hunk 18]
```diff
@@ -99,180 +115,477 @@ static bool is_transposed(const shape& s)
...
+        double first_time = -1;
+        // Initialize to default solution index
+        rocblas_int bestSol = 0;
+        for(auto sol : solution_indices)
+        {
+            // Warmup: the first few calls to an op. may not be representative since there is
+            // more time taken initializing caches, etc. so we won't time them.
+            for(rocblas_int cc = 0; cc < cold_calls; ++cc)
+            {
+                run(ctx, input_args, sol);
```
Comment (pfultz2) [inline src/targets/gpu/gemm_impl.cpp:431]: Just call this once, dont use a loop for cold calls. Also use `run_func` not `run`.

File: src/targets/gpu/gemm_impl.cpp [hunk 19]
```diff
@@ -99,180 +115,477 @@ static bool is_transposed(const shape& s)
...
+            }
+            double host_time = 0.0;
+            // Define the function to be timed
+            auto run_func = [&]() {
+                run(ctx, input_args, sol);
+                ctx.finish();
+            };
+            for(rocblas_int hc = 0; hc < hot_calls; ++hc)
```
Comment (pfultz2) [inline src/targets/gpu/gemm_impl.cpp:439]: Dont use `rocblas_int`

File: src/targets/gpu/gemm_impl.cpp [hunk 20]
```diff
@@ -99,180 +115,477 @@ static bool is_transposed(const shape& s)
...
+    void* beta_v  = nullptr;
+    flag_type int8_flag;
+    rocblas_int lda, ldb, ldc, ldd;
+    rocblas_int a_stride, b_stride, c_stride, d_stride;
+    rocblas_datatype compute_type, arg_type, output_type;
+    bool strided_batched = true, is_3inputs = true, compute_fp32 = true;
+#ifdef ROCBLAS_BETA_FEATURES_API
+    // tuning meta parameters
+    rocblas_int cold_calls = 18;
+    rocblas_int hot_calls  = 40;
```
Comment (pfultz2) [inline src/targets/gpu/gemm_impl.cpp:481]: This should be defined as a constant in the `tune` method.

File: src/targets/gpu/gemm_impl.cpp [hunk 21]
```diff
@@ -99,180 +115,477 @@ static bool is_transposed(const shape& s)
...
+            host_time /= hot_calls;
+
+            // debug only: track time for first solution.
+            if(first_time < 0)
+                first_time = host_time;
+
+            // track current best
+            if(host_time < bestTime)
+            {
+                printf(" current best index  %d, time %g\n", sol, host_time);
```
Comment (pfultz2) [inline src/targets/gpu/gemm_impl.cpp:456]: Use `std::cout`.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 22]
```diff
@@ -24,28 +24,75 @@
...
 #define MIGRAPHX_GUARD_RTGLIB_GEMM_IMPL_HPP
 
+#include <iterator>
 #include <migraphx/shape.hpp>
 #include <migraphx/argument.hpp>
+#include <migraphx/generate.hpp>
 #include <migraphx/gpu/context.hpp>
+#include <migraphx/reduce_dims.hpp>
+#include <migraphx/gpu/hip.hpp>
+#include <migraphx/time.hpp>
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:34]: These extra headers can be removed or moved to the .cpp file.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 23]
```diff
@@ -24,28 +24,75 @@
...
+
+using milliseconds = std::chrono::duration<double, std::milli>;
+using microseconds = std::chrono::duration<double, std::micro>;
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:37]: These can be removed from the header.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp [hunk 24]
```diff
@@ -24,28 +24,75 @@
...
-          const std::vector<argument>& args,
-          int32_t alpha,
-          int32_t beta,
-          bool int8_x4_format,
-          bool compute_fp32);
+#if ROCBLAS_VERSION_MAJOR >= 2 && ROCBLAS_VERSION_MINOR >= 38
+using flag_type = rocblas_gemm_flags;
+#else
+using flag_type = int;
+#endif
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:47]: This can be removed from the header and moved to cpp file.

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 25]
```diff
@@ -24,6 +24,8 @@
 #ifndef MIGRAPHX_GUARD_RTGLIB_GPU_GEMM_HPP
 #define MIGRAPHX_GUARD_RTGLIB_GPU_GEMM_HPP
 
+#include <algorithm>
+#include <iterator>
```
Comment (pfultz2) [inline src/targets/gpu/include/migraphx/gpu/gemm.hpp:28]: These headers can be removed.

File: .github/workflows/add-to-project.yaml [hunk 26]
```diff
@@ -15,5 +15,5 @@ jobs:
     steps:
       - uses: actions/add-to-project@v0.4.0
         with:
-          project-url: https://github.com/orgs/ROCmSoftwarePlatform/projects/26
+          project-url: https://github.com/orgs/ROCmSoftwarePlatform/projects/20
```
Comment (umangyadav) [inline .github/workflows/add-to-project.yaml:18]: Is this change necessary ? undo this change otherwise.

File: .github/workflows/ci.yaml [hunk 27]
```diff
@@ -39,8 +39,10 @@ jobs:
 
     - name: Prepare timestamp
       id: cache_timestamp
-      shell: bash 
-      run: echo timestamp="$(date +'%Y-%m-%dT%H:%M:%S')" >> $GITHUB_OUTPUT
+      shell: cmake -P {0}
```
Comment (umangyadav) [inline .github/workflows/ci.yaml:42]: You need to rebase with respect to develop properly.

File: Jenkinsfile [hunk 28]
```diff
@@ -28,7 +28,7 @@ def rocmtestnode(Map conf) {
             rm -rf build
             mkdir build
             cd build
-            cmake -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DBUILD_DEV=On ${flags} ..
+            cmake -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache ${flags} ..
```
Comment (umangyadav) [inline Jenkinsfile:31]: Undo this change as well.

File: install_deps.cmake [hunk 29]
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/cmake -P
```
Comment (umangyadav) [inline install_deps.cmake:1]: This file was deleted in develop .

**Review by bpickrel — COMMENTED** [review]
@pfultz2 this PR code no longer checks the status returned by rocblas_invoke().  Do we need to retain the ability to check for degraded performance, etc?

File: install_deps.cmake (hunk 29 above)
Comment (bpickrel) [reply install_deps.cmake]: ok; deleted

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp (hunk 25 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: done

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 24 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:47]: done

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 23 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:37]: done

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 22 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp:34]: done

File: src/targets/gpu/gemm_impl.cpp (hunk 21 above)
Comment (bpickrel) [reply src/targets/gpu/gemm_impl.cpp]: done.  Not sure if the message is needed at all.

File: src/targets/gpu/gemm_impl.cpp (hunk 20 above)
Comment (bpickrel) [reply src/targets/gpu/gemm_impl.cpp]: done

File: src/targets/gpu/gemm_impl.cpp (hunk 19 above)
Comment (bpickrel) [reply src/targets/gpu/gemm_impl.cpp]: changed to int

File: src/targets/gpu/gemm_impl.cpp (hunk 18 above)
Comment (bpickrel) [reply src/targets/gpu/gemm_impl.cpp:440]: Changing the parameter to 1 but leaving the loop in place, for now.

File: src/targets/gpu/gemm_impl.cpp [hunk 30]
```diff
@@ -99,180 +115,477 @@ static bool is_transposed(const shape& s)
...
+        //
+        rocblas_int list_size = 0;
+        std::vector<rocblas_int> solution_indices;
+        if(strided_batched)
+        {
+            auto common_args = create_strided_batched_args_common(ctx, input_args);
+            auto ded_args = pack(rocblas_gemm_algo_solution_index, int8_flag, nullptr, &list_size);
+
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex_get_solutions,
+                           pack_join(common_args, ded_args));
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/targets/gpu/gemm_impl.cpp]: done

File: src/targets/gpu/CMakeLists.txt (hunk 17 above)
Comment (bpickrel) [reply src/targets/gpu/CMakeLists.txt]: changed

File: src/targets/gpu/include/migraphx/gpu/convolution.hpp [hunk 31]
```diff
@@ -163,6 +167,20 @@ struct miopen_convolution
         auto w_desc                = make_tensor(reshape_if_1d(inputs[1]), int8_x4_format);
         auto y_desc                = make_tensor(reshape_if_1d(output_shape));
         std::size_t workspace_size = 0;
+
+        // TODO:  Using an environment variable to disable convolution tuning while still allowing
+        // other (i.e. GEMM) tuning with --exhaustive-tune, is a workaround for a problem in which
+        // convolution tuning takes multiple hours.
+        const bool convo_tune_enabled = enabled(MIGRAPHX_ENABLE_CONV_TUNING{});
+        if(ctx.get_exhaustive_tune_flag() and not convo_tune_enabled)
+        {
```
Comment (bpickrel) [inline src/targets/gpu/include/migraphx/gpu/convolution.hpp:176]: This causes a flood of warning messages.  Do we have a tidy way to limit this message to only output once per compilation?

File: .github/workflows/add-to-project.yaml (hunk 26 above)
Comment (bpickrel) [reply .github/workflows/add-to-project.yaml]: merge error; fixed

File: .github/workflows/ci.yaml (hunk 27 above)
Comment (bpickrel) [reply .github/workflows/ci.yaml]: merge error; fixed

File: Jenkinsfile (hunk 28 above)
Comment (bpickrel) [reply Jenkinsfile]: fixed

File: src/targets/gpu/include/migraphx/gpu/gemm.hpp [hunk 32]
```diff
@@ -111,19 +110,21 @@ struct rocblas_gemm
...
     {
+        std::vector<shape> input_shapes;
+        std::transform(args.begin(),
+                       args.end(),
+                       std::back_inserter(input_shapes),
+                       [](const argument& x) { return x.get_shape(); });
         if(this->name() == "gpu::gemm")
         {
-            gemm(ctx, output_shape, args, alpha, beta, int8_x4_format, compute_fp32);
+            gemm_impl<float>(output_shape, input_shapes, alpha, beta, int8_x4_format, compute_fp32)
```
[the remark this answers is not in the document]
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm.hpp]: at long last, worked out a way to implement a visit lambda with the refactored code and the `half` verification test now passes.

File: src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp (hunk 16 above)
Comment (bpickrel) [reply src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp]: resolved it by returning 0 in the bad case (should only happen when get_batch_stride isn't needed)

**Comment by bpickrel** [comment]
Closed because of corrupted merge; replaced by PR 1668
